### DISCUSSION TITLE
'linkerd routes' command doesn't show anything when running with low value of time window option

### DISCUSSION THREAD

**Discussion by nfsp3k** [linkerd/linkerd2#4829] [General]
`linkerd routes` doesn't show anything when I set the value of `-t` option relatively low.
For example, a load generator sends 300RPS workload and the output of `linkerd routes` is as follows when the value of `-t` option is:
'10s'
![image](https://user-images.githubusercontent.com/4262433/89127024-58e30280-d525-11ea-9164-7503f0f99541.png)

'15s' (RPS is unstable and varies.)
![image](https://user-images.githubusercontent.com/4262433/89127031-626c6a80-d525-11ea-89a7-8d5a47a600aa.png)

'20s'
![image](https://user-images.githubusercontent.com/4262433/89127061-9b0c4400-d525-11ea-8c59-5dea593ea0f4.png)

'30s'
![image](https://user-images.githubusercontent.com/4262433/89127067-a65f6f80-d525-11ea-89ba-17c5f650ecfe.png)

I configured Linkerd-Prometheus to scrape metrics every second as follows:
![image](https://user-images.githubusercontent.com/4262433/89127085-d6a70e00-d525-11ea-9913-705a7f358f02.png)
![image](https://user-images.githubusercontent.com/4262433/89127098-f76f6380-d525-11ea-822b-a22fde89caaf.png)

Actually, `linkerd routes` works perfectly with the low value of `-t` option (e.g., 2s) some days ago, but it doesn't work correctly now and I don't know the reason.

The current environment of mine is as follows:
  - Linkerd version: v2.8.1
  - K8S version: v1.17.9
  - CNI: Calico
  - Other components deployed together: Jaeger, Prometheus-adapter, Linkerd Service Profile

I think it would be a poor question but I am new to Linkerd and I hope you understand.
Thanks in advance.

**Comment by cpretzer** [comment]
Hi @nfsp3k the `-t` flag specifies the `time window` that the `routes` command uses when it generates the prometheus query to show the metrics. You can see the prometheus queries that are generated by setting `-log-level=debug` in the `linkerd-controller` deployment.

Here's an example that I generated from our emojivoto application:

`histogram_quantile(0.95, sum(irate(route_response_latency_ms_bucket{direction="inbound", dst=~"(web-svc.emojivoto.svc.cluster.local)(:\\d+)?", namespace="emojivoto"}[10s])) by (le, dst, rt_route))`

The query uses the value specified by `-t` in the prometheus [irate](https://prometheus.io/docs/prometheus/latest/querying/functions/#irate) function, so the query is for the `route_response_latency_ms_bucket` last two data points looking 10 seconds back. It's entirely possible that there are no data points, which is why you are seeing no values.

I ran `watch linkerd routes svc/web-svc -t 10s` and eventually some results would display. If you get the queries from the `-log-level=debug` output of the `linkerd-controller` you can access the prometheus query UI with `kubectl port-forward svc/linkerd-prometheus -n linkerd 9090` and run the queries there to see what is displayed.
